torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,701 @@
+
+        corr = torch.einsum("aijk,aijh->ajkh", fmap1, fmap2)
+        corr = corr.view(batch_size, h, w, 1, w)
+        corr_volume = corr / torch.sqrt(torch.tensor(num_channels, device=corr.device))
+
+        corr_volume = corr_volume.reshape(batch_size * h * w, 1, 1, w)
+        corr_pyramid = [corr_volume]
+        for _ in range(self.num_levels - 1):
+            corr_volume = F.avg_pool2d(corr_volume, kernel_size=(1, 2), stride=(1, 2))
+            corr_pyramid.append(corr_volume)
+
+        return corr_pyramid
+
+
+class CorrBlock1d(nn.Module):
+    """The row-wise correlation block.
+
+    Use indexes from correlation pyramid to create correlation features.
+    The "indexing" of a given centroid pixel x' is done by concatenating its surrounding row neighbours
+    within radius
+    """
+
+    def __init__(self, *, num_levels: int = 4, radius: int = 4):
+        super().__init__()
+        self.radius = radius
+        self.out_channels = num_levels * (2 * radius + 1)
+
+    def forward(self, centroids_coords: Tensor, corr_pyramid: List[Tensor]) -> Tensor:
+        """Return correlation features by indexing from the pyramid."""
+        neighborhood_side_len = 2 * self.radius + 1  # see note in __init__ about out_channels
+        di = torch.linspace(-self.radius, self.radius, neighborhood_side_len, device=centroids_coords.device)
+        di = di.view(1, 1, neighborhood_side_len, 1).to(centroids_coords.device)
+
+        batch_size, _, h, w = centroids_coords.shape  # _ = 2 but we only use the first one
+        # We only consider 1d and take the first dim only
+        centroids_coords = centroids_coords[:, :1].permute(0, 2, 3, 1).reshape(batch_size * h * w, 1, 1, 1)
+
+        indexed_pyramid = []
+        for corr_volume in corr_pyramid:
+            x0 = centroids_coords + di  # end shape is (batch_size * h * w, 1, side_len, 1)
+            y0 = torch.zeros_like(x0)
+            sampling_coords = torch.cat([x0, y0], dim=-1)
+            indexed_corr_volume = grid_sample(corr_volume, sampling_coords, align_corners=True, mode="bilinear").view(
+                batch_size, h, w, -1
+            )
+            indexed_pyramid.append(indexed_corr_volume)
+            centroids_coords = centroids_coords / 2
+
+        corr_features = torch.cat(indexed_pyramid, dim=-1).permute(0, 3, 1, 2).contiguous()
+
+        expected_output_shape = (batch_size, self.out_channels, h, w)
+        torch._assert(
+            corr_features.shape == expected_output_shape,
+            f"Output shape of index pyramid is incorrect. Should be {expected_output_shape}, got {corr_features.shape}",
+        )
+        return corr_features
+
+
+class RaftStereo(nn.Module):
+    def __init__(
+        self,
+        *,
+        feature_encoder: FeatureEncoder,
+        context_encoder: MultiLevelContextEncoder,
+        corr_pyramid: CorrPyramid1d,
+        corr_block: CorrBlock1d,
+        update_block: MultiLevelUpdateBlock,
```

Comment:
@datumbox at the beginning I use `nn.Module`, but after that mypy will complain if I access the variable of the object. For instance if I set feature_encoder as `nn.Module`, I will get the following mypy error:
```
raft_stereo.py:395: note: Right operand is of type "Union[Tensor, Module]"
raft_stereo.py:402: error: Unsupported operand types for // ("int" and Module)  [operator]
                fmap1.shape[-2:] == (h // self.base_downsampling_ratio, w // self.base_downsampling_ratio),
```
it is because the `base_downsampling_ratio` comes from feature encoder:
```
        self.base_downsampling_ratio = feature_encoder.base_downsampling_ratio
```
and mypy assumed any class variable of `nn.Module` as type `Union[Tensor, Module]`.

I think in this case we need to make decision whether to ignore these mypy error, or use specific class (that have the variable we wanted)

Reply:
Yeah fair enough. Let's keep it as is for now and see if we want to overwrite or workaround mypy when we clean up the implementation before graduating from prototype.